02-household-queries/optimize_encoder.py

```diff
@@ -0,0 +1,249 @@
+    "BAAI/bge-small-en-v1.5": {
+        "func": SentenceTransformerEmbeddings(model_name="BAAI/bge-small-en-v1.5"),
+        "token_limit": 512,
+    },
+    "mixedbread-ai/mxbai-embed-large-v1": {
+        "func": SentenceTransformerEmbeddings(
+            model_name="mixedbread-ai/mxbai-embed-large-v1"
+        ),
+        "token_limit": 1024,
+    },
+}
+
+
+def load_training_json():
+    with open("question_answer_citations.json", encoding="utf-8") as data_file:
+        json_data = json.load(data_file)
+        # print(json.dumps(json_data, indent=2))
+        return json_data
+
+
+def compute_percent_retrieved(retrieved_cards, guru_cards):
+    missed_cards = set(guru_cards) - set(retrieved_cards)
+    return (len(guru_cards) - len(missed_cards)) / len(guru_cards)
+
+
+def count_extra_cards(retrieved_cards, guru_cards):
+    extra_cards = set(retrieved_cards) - set(guru_cards)
+    return len(extra_cards)
+
+
+# split text into chunks
+def get_text_chunks_langchain(
+    text, source, chunk_size, chunk_overlap, token_limit, text_splitter_choice
+):
+    if text_splitter_choice == "2":
+        text_splitter = NLTKTextSplitter()
+    elif text_splitter_choice == "3":
+        text_splitter = SpacyTextSplitter()
+    else:
+        text_splitter = RecursiveCharacterTextSplitter(
+            chunk_size=chunk_size, chunk_overlap=chunk_overlap
+        )
+
+    texts = text_splitter.split_text(source + "\n\n" + text)
+    # print("  Split into", len(texts))
+    for t in texts:
+        token_count = llm.get_num_tokens(t)
+        if token_count > token_limit:
+            print(f"Exceeded token limit of {token_limit}: {token_count}; {t}")
+
+    docs = [
+        Document(page_content=t, metadata={"source": source.strip()}) for t in texts
+    ]
+
+    return docs
+
+
+# Chunk the json data and load into vector db
+def add_json_html_data_to_vector_db(
+    vectordb,
+    file_path,
+    content_key,
+    index_key,
+    chunk_size,
+    chunk_overlap,
+    token_limit,
+    text_splitter_choice,
+):
+    data_file = open(file_path, encoding="utf-8")
+    json_data = json.load(data_file)
+    for content in json_data:
+        if not content[index_key].strip().endswith("?"):
+            continue
+        soup = BeautifulSoup(content[content_key], "html.parser")
+        text = soup.get_text(separator="\n", strip=True)
+        # print("Processing document:", content[index_key])
+        chunks = get_text_chunks_langchain(
+            text,
+            content[index_key],
+            chunk_size=chunk_size,
+            chunk_overlap=chunk_overlap,
+            token_limit=token_limit,
+            text_splitter_choice=text_splitter_choice,
+        )
+        vectordb.add_documents(documents=chunks)
+
+
+def ingest_call(vectordb, chunk_size, chunk_overlap, token_limit, text_splitter_choice):
+    # download from https://drive.google.com/drive/folders/1DkAQ03bBVIPoO1d8gcHVnilQ-9VXfhJ8?usp=drive_link
```

Comment:
I think you'd need to adjust token_limit to take into account the chunk_size, right? E.g., for a token_limit of `256`, it's OK for the chunk size to be `368` if the overlap is `128` because the the last 128 of the chunk will be included in the next chunk (where it will be the first 128 tokens, so still within the limit)

Doesn't matter for our experiment here since we retrieve the whole card if any of the chunks match, but for the `tanf.pdf` it would have an impact for example